candid_diff/design/candiff-CLI-tool.md

```diff
@@ -0,0 +1,89 @@
+# Tool design (`candiff`)
+
+The `candid_diff` package exposes the `candiff` tool for scripting Candid value comparisons that have human-readable error messages.
+
+The tool compares input and output canister messages to those recorded earlier, from earlier run(s).
+
+## Use cases
+
+- General-purpose debugging / development tool for Candid data
+- In CI scripts: summarize deviations from expected responses
+
+## Candiff input format
+
+Accept binary and text options for Candid data to compare:
+
+- raw candid (what `raw` means for `dfx` input/output)
+- candid text format
+
+## Candiff output format
+ 
+- No default output when diff is empty
+- Human-readable default output when diff is non-empty
+
+The human-readable diff answers the question:
+
+> How do I structurally edit the first value into the second, with the fewest ("small") edits?
+
+## Data and type changes
+
+Rather than compare two values textually, the tool is structural, like Candid itself.
+
+The tool does not attempt to compare candid values that do not share a
+common type, but we do not know if this the case or not, a priori.
+When provided, this common type helps inform the output information,
+but it is not required.  When absent, the tool gives less helpful
+information on values without a common type.
+
+### Typing
+
+Two cases:
+
+- Type invariant: Both values have a common type.
```

Comment:
Do we want to consider subtyping, e.g. two values having a common subtype?

Reply:
Maybe?  Or "upgrades", if those are distinct notions (?).

The simplest use case doesn't require that, but we should note that other (future) uses could implement more of the type theory of candid, including that relation, and follow it.

Reply:
Coming back to this question now with some more implementation and testing experience.

Currently, I don't see that `check_type` even supports the subtyping relation.  Does it?

Reply:
If not, that's okay.  Perhaps we can revisit this question once the Candid crate itself supports subtyping?

Reply:
No, subtyping is partially supported in serialization. You need two types/two candid files for subtyping to make sense. `check_type` only checks for a single candid file.